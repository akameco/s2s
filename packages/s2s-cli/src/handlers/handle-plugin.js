// @flow
import path from 'path'
import fs from 'fs'
import type { Path, AfterHook, Plugin, Handler } from 'types'
import {
  getOutputPath,
  writeFileSync,
  relativeFromCwd,
  resolveInputPath,
} from '../utils'
import runHooks from '../hooks'
import { formatText } from '../reporters'

type Opts = {
  eventPath: Path,
  plugin: Plugin,
  hooks: Array<AfterHook>,
}

export function handlePlugin(
  handler: Handler,
  { eventPath, plugin, hooks = [] }: Opts
) {
  const filename = resolveInputPath(plugin.input, eventPath)
  const content = fs.readFileSync(filename, 'utf8')

  const handlerResult = handler(content, { eventPath, plugin, filename })

  const { code, meta } =
    handlerResult && typeof handlerResult === 'string'
      ? { code: handlerResult, meta: { handlerName: 'S2S' } }
      : handlerResult

  if (!code && code === '') {
    return
  }

  const outputFilename = resolveInputPath(plugin.output, eventPath)
  const result = runHooks(code, outputFilename, hooks)

  const outputPath = plugin.output
    ? getOutputPath(
        plugin.output.replace('[name]', path.parse(eventPath).name),
        eventPath
      )
    : eventPath

  writeFileSync(outputPath, result)

  const { handlerName } = meta

  const outputPrefix =
    handlerName +
    (meta.pluginName && meta.pluginName !== '' ? `:${meta.pluginName}` : '')

  console.log(formatText(outputPrefix, relativeFromCwd(eventPath), outputPath))
}
