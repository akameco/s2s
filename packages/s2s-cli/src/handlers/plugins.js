// @flow
import path from 'path'
import fs from 'fs'
import KeyLocker from 'key-locker'
import defaultHanlder from 's2s-handler-babel'
import micromatch from 'micromatch'
import type { Path, AfterHook, Plugin, EventType, HanlderFunc } from 'types'
import {
  getOutputPath,
  writeFileSync,
  toErrorStack,
  log,
  relativeFromCwd,
  resolveInputPath,
} from '../utils'
import runHooks from '../hooks'
import { formatText } from '../reporters/'

type Opts = {
  eventPath: Path,
  plugin: Plugin,
  hooks: AfterHook[],
}

export function handlePlugin(
  handler: HanlderFunc,
  { eventPath, plugin, hooks = [] }: Opts
) {
  const filename = resolveInputPath(plugin.input, eventPath)
  const content = fs.readFileSync(filename, 'utf8')

  const code = handler(content, { eventPath, plugin, filename })

  if (!code && code === '') {
    return
  }

  const result = runHooks(filename, code, hooks)

  const outputPath = plugin.output
    ? getOutputPath(
        plugin.output.replace('[name]', path.parse(eventPath).name),
        eventPath
      )
    : eventPath

  writeFileSync(outputPath, result)
  log(formatText('S2S', relativeFromCwd(eventPath), outputPath))
}

function validate(
  eventPath: Path,
  eventType: EventType,
  plugin: Plugin
): boolean {
  if (typeof plugin.test === 'string' || Array.isArray(plugin.test)) {
    return !micromatch.some(eventPath, plugin.test)
  } else if (!plugin.test.test(eventPath)) {
    return false
  }

  if (plugin.only && !plugin.only.includes(eventType)) {
    return false
  }

  return true
}

const lock = new KeyLocker()

export default function handlePlugins(
  eventPath: Path,
  eventType: EventType,
  plugins: Plugin[] = [],
  hooks: AfterHook[] = []
) {
  if (lock.has(eventPath)) {
    return
  }

  for (const plugin of plugins) {
    try {
      if (!validate(eventPath, eventType, plugin)) {
        continue // eslint-disable-line
      }

      lock.add(eventPath)

      const handler = plugin.handler ? plugin.handler : defaultHanlder
      handlePlugin(handler, { eventPath, plugin, hooks })
    } catch (err) {
      console.error(toErrorStack(err))
    }
  }
}
