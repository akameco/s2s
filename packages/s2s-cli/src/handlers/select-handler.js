// @flow
import micromatch from 'micromatch'
import handlerBabel from 's2s-handler-babel'
import hanlderTypeScript from 's2s-handler-typescript'
import type { Path, Handler } from 'types'

const DEFAULT_HANDLE_MAPPER = {
  '*.(js|jsx)': handlerBabel,
  '*.(ts|tsx)': hanlderTypeScript,
}

type HandlerMapper = { [extensions: string]: Handler }

export function selectHandler(
  handlerMapper: HandlerMapper,
  handler?: Handler,
  filepath: Path
): Handler {
  if (handler) {
    return handler
  }

  const finalHandlerMapper =
    Object.keys(handlerMapper).length > 0
      ? handlerMapper
      : DEFAULT_HANDLE_MAPPER

  for (const key of Object.keys(finalHandlerMapper)) {
    if (micromatch.isMatch(filepath, key, { matchBase: true })) {
      return finalHandlerMapper[key]
    }
  }

  throw new Error('any handlers not match')
}
