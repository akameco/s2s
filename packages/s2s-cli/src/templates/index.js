// @flow
import path from 'path'
import fs from 'fs'
import chalk from 'chalk'
import ejs from 'ejs'
import slash from 'slash'
import type { Path, Template } from 'types'
import { formatText, trimAndFormatPath } from '../reporters'
import { isAlreadyExist, relativeFromCwd, getOutputPath } from '../utils'
import some from '../utils/some'

type CopyError = Error & { path: string, code: string }

function handleCopyError(err: CopyError): void {
  if (err.name === 'CpFileError' && err.code === 'ENOENT') {
    const errorText = `${chalk.reset.inverse.bold.red(
      'TEMPLATE'
    )} ${trimAndFormatPath(err.path)}
          no such file or directory`
    console.log(errorText)
  } else {
    console.log(err.stack)
  }
}

const DEFAULT_TEMPLATES_DIR = 'templates'

function validate(template: Template, eventPath: Path) {
  if (typeof template.test === 'string' || Array.isArray(template.test)) {
    return some(eventPath, template.test)
  } else if (!template.test.test(eventPath)) {
    return false
  }

  return true
}

function handleTemplate(
  eventPath: Path,
  template: Template,
  templatesDir: string
) {
  if (!validate(template, eventPath)) {
    return
  }

  const outputPath = template.output
    ? getOutputPath(template.output, eventPath)
    : eventPath

  if (isAlreadyExist(outputPath)) {
    return
  }

  const templatePath = path.join(templatesDir, template.input)
  const str = fs.readFileSync(templatePath, 'utf8')

  const result = ejs.render(str, {
    FILENAME: path.parse(outputPath).name,
    DIRNAME: path.basename(path.dirname(outputPath)),
  })

  fs.writeFileSync(outputPath, result)

  console.log(
    formatText('TEMPLATE', relativeFromCwd(templatePath), slash(outputPath))
  )
}

export default function handleTemplates(
  eventPath: Path,
  templates: Template[] = [],
  templatesDir: string = DEFAULT_TEMPLATES_DIR
) {
  for (const template of templates) {
    try {
      handleTemplate(eventPath, template, templatesDir)
    } catch (err) {
      handleCopyError(err)
    }
  }
}
