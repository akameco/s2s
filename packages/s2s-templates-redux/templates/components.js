// @flow
import * as React from 'react'

type Props = {}

class Cart extends React.Component<Props> {
  render() {
    return <div>hello</div>
  }
}

export default Cart
