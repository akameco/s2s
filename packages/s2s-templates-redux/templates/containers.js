// @flow
import * as React from 'react'
import { connect, type Connector } from 'react-redux'
import type { State as RootState } from '../../types'
import * as selectors from './selectors'

type Props = {}

class Container extends React.Component<Props> {
  render() {
    const {} = this.props
    return <div>{}</div>
  }
}

const ms = (state: RootState) => ({
  TODO: selectors.TODO,
})

const conn: Connector<{}, Props> = connect(ms, { TODO })

export default conn(Container)
