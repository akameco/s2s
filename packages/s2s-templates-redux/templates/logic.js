// @flow
import type { ThunkAction } from '../../types'
import * as actions from './actions'
