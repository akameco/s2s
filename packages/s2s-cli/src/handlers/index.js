// @flow
import KeyLocker from 'key-locker'
import type { Path, Plugin, EventType, Config } from 'types'
import { toErrorStack } from '../utils'
import some from '../utils/some'
import { handlePlugin } from './handle-plugin'
import { selectHandler } from './select-handler'

function validate(plugin: Plugin, eventPath: Path, eventType: EventType) {
  if (typeof plugin.test === 'string' || Array.isArray(plugin.test)) {
    return some(eventPath, plugin.test)
  } else if (!plugin.test.test(eventPath)) {
    return false
  }

  if (plugin.only && !plugin.only.includes(eventType)) {
    return false
  }

  return true
}

const lock = new KeyLocker()

export default function handlePlugins(
  eventPath: Path,
  eventType: EventType,
  config: $Shape<Config> = {}
) {
  if (lock.has(eventPath)) {
    return
  }

  const { plugins = [], afterHooks: hooks = [], handlerMapper = {} } = config

  for (const plugin of plugins) {
    if (validate(plugin, eventPath, eventType)) {
      lock.add(eventPath)

      try {
        const handler = selectHandler(handlerMapper, plugin.handler, eventPath)
        handlePlugin(handler, { eventPath, plugin, hooks })
      } catch (error) {
        if (error.code === 'ENOENT') {
          console.error(error.message)
        } else {
          console.error(toErrorStack(error))
        }
      }
    }
  }
}
