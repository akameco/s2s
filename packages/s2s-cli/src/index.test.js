// @flow
import cases from 'jest-in-case'
import type { Config } from 'types'
import * as plugins from './handlers'
import * as templates from './templates'
import m from '.'

let logSpy
let handlePluginsSpy
let handleTemplateSpy
let watcher

function setup(options = {}): Config {
  return {
    watch: 'app',
    plugins: [{ test: /dummy/, plugin: 'dummy' }],
    templates: [{ test: /dummy/, input: 'dummy' }],
    ...options,
  }
}

beforeEach(() => {
  logSpy = jest.spyOn(console, 'log').mockImplementation(x => x)
  handlePluginsSpy = jest.spyOn(plugins, 'default')
  handleTemplateSpy = jest.spyOn(templates, 'default')
})

afterEach(() => {
  logSpy.mockRestore()
  handlePluginsSpy.mockRestore()
  handleTemplateSpy.mockRestore()
  if (watcher && watcher.close) {
    watcher.close()
  }
})

cases(
  'throw errors',
  options => {
    expect(() => {
      m(setup(options.input))
    }).toThrowErrorMatchingSnapshot()
  },
  [
    // $FlowFixMe
    { name: 'opts.watch == null', input: { watch: null } },
    // $FlowFixMe
    { name: 'opts.plugins !== Array', input: { plugins: 'not array' } },
    // $FlowFixMe
    { name: 'opts.templates !== Array', input: { templates: 'string' } },
    { name: 'opts.after !== Array', input: { afterHooks: 'string' } },
  ]
)

test('Displayed "start" on the Consle', () => {
  watcher = m(setup())
  expect(logSpy.mock.calls).toHaveLength(1)
  expect(logSpy.mock.calls[0][0]).toMatch('start')
})

test('pluginsとtemplatesが渡されないとき', () => {
  const options = setup()
  delete options.plugins
  delete options.templates
  watcher = m(options)
  expect(logSpy.mock.calls).toHaveLength(1)
  expect(logSpy.mock.calls[0][0]).toMatch('start')
})

test('ファイルが追加されたときhandlePluginsが呼ばれる', () => {
  watcher = m(setup())
  watcher.emit('add', 'hello')
  const result = handlePluginsSpy.mock.calls[0]
  expect(result[0]).toBe('hello')
  expect(result[1]).toBe('add')
  expect(result[2]).toMatchSnapshot()
})

test('pluginsがない場合、ファイルが追加されたときhandlePluginsは呼ばれない', () => {
  watcher = m(setup({ plugins: [] }))
  watcher.emit('add', 'hello')
  expect(handlePluginsSpy).not.toHaveBeenCalled()
})

test('ファイルが追加されたときhandleTemplatesが呼ばれる', () => {
  watcher = m(setup())
  watcher.emit('add', 'hello')
  const result = handleTemplateSpy.mock.calls[0]
  expect(result).toMatchInlineSnapshot(`
    Array [
      "hello",
      Array [
        Object {
          "input": "dummy",
          "test": /dummy/,
        },
      ],
      undefined,
    ]
  `)
})

test('templatesがない場合、ファイルが追加されたときhandleTemplatesは呼ばれない', () => {
  watcher = m(setup({ templates: [] }))
  watcher.emit('add', 'hello')
  expect(handleTemplateSpy).not.toHaveBeenCalled()
})

test('prettier = falseのとき、 afterHooksが空配列でhandlePluginが呼ばれる', () => {
  watcher = m(setup({ prettier: false }))
  watcher.emit('add', 'hello')
  const result = handlePluginsSpy.mock.calls[0]
  expect(result[2].afterHooks).toStrictEqual([])
})

test('ignoredに指定されたパターンに該当するファイルは、呼ばれない', () => {
  watcher = m(setup({ ignored: ['hoge'] }))
  expect(watcher.options.ignored).toStrictEqual(['hoge'])
})
