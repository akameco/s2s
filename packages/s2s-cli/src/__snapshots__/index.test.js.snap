// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throw errors opts.after !== Array 1`] = `"Expected a Array got string"`;

exports[`throw errors opts.plugins !== Array 1`] = `"Expected a Array, got string"`;

exports[`throw errors opts.templates !== Array 1`] = `"Expected a Array got string"`;

exports[`throw errors opts.watch == null 1`] = `"required watch"`;

exports[`ファイルが追加されたときhandlePluginsが呼ばれる 1`] = `
Object {
  "afterHooks": Array [
    [Function],
  ],
  "ignored": Array [
    ".git/",
    "node_modules/",
  ],
  "plugins": Array [
    Object {
      "plugin": "dummy",
      "test": /dummy/,
    },
  ],
  "prettier": true,
  "templates": Array [
    Object {
      "input": "dummy",
      "test": /dummy/,
    },
  ],
  "watch": "app",
}
`;
